<a href="https://colab.research.google.com/github/manugoyal1/azure-static-web-app-angular-basic/blob/main/OCR_Classification_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# import libraries.
!pip install imbalanced-learn
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, roc_auc_score
import time
from skimage.transform import resize
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import roc_auc_score
import pandas as pd
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve
from sklearn.model_selection import GridSearchCV
!pip install tensorflow
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
import tensorflow as tf
from sklearn.metrics import roc_auc_score, roc_curve
import matplotlib.pyplot as plt

### Load data
##file_path = 'your_data_file.csv'  # Update this path to your actual data file location
##data = pd.read_csv(file_path)

### Separate features and labels
##X = data.drop('labels', axis=1)
##y = data['labels']


In [3]:
from google.colab import files

# Optional: You can skip this step if you are running the code on your own
# machine
uploaded = files.upload()

Saving JCB704_MNIST_resized.csv to JCB704_MNIST_resized.csv


In [4]:
data = pd.read_csv('JCB704_MNIST_resized.csv')
# Begin by familiarizing yourself with the dataset.
print("First five rows of the training data:")
print(data.head())



First five rows of the training data:
     0    1    2    3    4    5    6    7    8    9  ...       187       188  \
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000000  0.000000   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000000  0.000000   
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000123  0.000073   
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000002  0.010165   
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000000  0.000000   

        189           190  191  192  193  194  195  labels  
0  0.000000  0.000000e+00  0.0  0.0  0.0  0.0  0.0       1  
1  0.000000  0.000000e+00  0.0  0.0  0.0  0.0  0.0       1  
2  0.000014  2.783140e-07  0.0  0.0  0.0  0.0  0.0       1  
3  0.033733  1.094800e-03  0.0  0.0  0.0  0.0  0.0       0  
4  0.000000  0.000000e+00  0.0  0.0  0.0  0.0  0.0       0  

[5 rows x 197 columns]


In [5]:
# Separate features and labels
X = data.drop('labels', axis=1)
y = data['labels']

In [6]:
# Splitting data (Using only 10% for training as specified)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.1, random_state=42)


In [33]:
#Import the datetime module
import datetime
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, roc_auc_score

# Define the neural network model for the finalized training with 10% of the dataset
mlp = MLPClassifier(hidden_layer_sizes=(128, 64), activation='relu', solver='adam', max_iter=500, random_state=42)

# Split the dataset to use 10% for training
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.80, random_state=42, stratify=y)

# Start the training session timer
training_start_time = datetime.datetime.now()

# Training the model
mlp.fit(X_train, y_train)

# Evaluate the model
predictions = mlp.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
precision, recall, _, _ = precision_recall_fscore_support(y_test, predictions, average='macro')
auc = roc_auc_score(y_test, mlp.predict_proba(X_test)[:, 1])

# End the training session timer
training_end_time = datetime.datetime.now()

# Display the timing and performance metrics
# prompt: print results in single row with labels. use 4 decimal places for floats. use hh:mm:ss for datetime

print(f"Accuracy: {accuracy:.4f}, Precision: {precision:.4f}, Recall: {recall:.4f}, AUC: {auc:.4f}, Training Time: {(training_end_time - training_start_time)}")


Accuracy: 0.9885, Precision: 0.9886, Recall: 0.9885, AUC: 0.9992, Training Time: 0:00:16.092238



10%   Accuracy: 0.9664, Precision: 0.9665, Recall: 0.9664, AUC: 0.9938, Training Time: 0:00:01.859822

20%   Accuracy: 0.9769, Precision: 0.9770, Recall: 0.9769, AUC: 0.9971, Training Time: 0:00:05.270882

40%
Accuracy: 0.9786, Precision: 0.9787, Recall: 0.9786, AUC: 0.9978, Training Time: 0:00:05.887880

60% Accuracy: 0.9827, Precision: 0.9827, Recall: 0.9827, AUC: 0.9984, Training Time: 0:00:07.719535

80% Accuracy: 0.9885, Precision: 0.9886, Recall: 0.9885, AUC: 0.9992, Training Time: 0:00:10.560345


500 iterations

80%
 Accuracy: 0.9885, Precision: 0.9886, Recall: 0.9885, AUC: 0.9992, Training Time: 0:00:16.092238


In [26]:
# prompt: print results in single row with labels. use 4 decimal places for floats. use hh:mm:ss for datetime

print(f"Accuracy: {accuracy:.4f}, Precision: {precision:.4f}, Recall: {recall:.4f}, AUC: {auc:.4f}, Training Time: {(training_end_time - training_start_time)}")


Training Start Time: 2024-09-26 02:22:59.622874, Training End Time: 2024-09-26 02:23:04.600418, Accuracy: 0.9663865546218487, Precision: 0.9664771668020238, Recall: 0.9663865546218486, AUC: 0.9938114168577463


In [18]:
# Function to create and train a neural network model
def train_and_evaluate(max_iter):
    start_time = time.time()
    model = MLPClassifier(hidden_layer_sizes=(128, 64), activation='relu', solver='adam', max_iter=100, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test).ravel()
    training_time = time.time() - start_time

    # Evaluation metrics
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='macro', zero_division=0)
    recall = recall_score(y_test, y_pred, average='macro')
    auc = roc_auc_score(y_test, model.predict_proba(X_test), multi_class='ovr')

    print ('max_iter :', max_iter, 'accuracy: ', accuracy, 'precision: ', precision, 'recall: ', recall, 'auc: ', auc, 'training_time: ', training_time)
    #return {'max_iter': max_iter, 'accuracy': accuracy, 'precision': precision, 'recall': recall, 'auc': auc, 'training_time': training_time}

# Iterations to test
##iterations = [100, 200]
##results = [train_and_evaluate(iter) for iter in iterations]
results = train_and_evaluate(iter)

### Printing and plotting results
##print(results)
##for metric in ['accuracy', 'precision', 'recall', 'auc', 'training_time']:
##    plt.figure()
##    plt.plot([result['max_iter'] for result in results], [result[metric] for result in results], marker='o')
##    plt.title(f'{metric} vs. Max Iterations')
##    plt.xlabel('Max Iterations')
##    plt.ylabel(metric)
##    plt.grid(True)
##    plt.show()


ValueError: y should be a 1d array, got an array of shape (9758, 2) instead.